File: dart/ttagy/test/conformance_test.dart
import 'dart:io';
import 'package:test/test.dart';
import 'package:ttagy/ttagy.dart';

void main() {
  setUpAll(() {
    final root = Directory.current.parent.parent.path;
    final mockPath = '$root/target/debug/mock-agy';
    Platform.environment['AGY_PATH'] = mockPath;
  });

  test('Dart: extractStructuredJson handles markdown fences', () {
    const raw = 'Output:\n```json\n{\n  "status": "ok",\n  "count": 42\n}\n```\nDone';
    final extracted = extractStructuredJson(raw);
    expect(extracted, contains('"status": "ok"'));
  });

  test('Dart CTS: stream_normal parity', () async {
    final client = TtagyClient(autoFallback: true);
    final events = <TtagyStreamEvent>[];
    final contentBuf = StringBuffer();

    await for (final ev in client.streamChat(const TtagyRequest(
      prompt: 'scenario:stream_normal',
      model: 'gemini-3.7-flash',
      effort: 'low',
    ))) {
      events.add(ev);
      if (ev is TtagyContentDeltaEvent) {
        contentBuf.write(ev.textDelta);
      }
    }

    final types = events.map((e) => e.type).toList();
    expect(types, containsAllInOrder(['agy:init', 'agy:thinking_delta', 'agy:content_delta', 'agy:done']));
    expect(contentBuf.toString(), contains('Antigravity AI 助手'));
  });

  test('Dart CTS: structured_json parity', () async {
    final client = TtagyClient(autoFallback: true);
    final result = await client.runJson(
      const TtagyRequest(
        prompt: 'scenario:structured_json',
        model: 'gemini-3.7-flash',
        effort: 'low',
      ),
      (json) => json as Map<String, dynamic>,
    );

    expect(result['status'], equals('success'));
    expect(result['task'], equals('compression'));
    expect(result['files_count'], equals(3));
  });

  test('Dart CTS: quota_error parity', () async {
    final client = TtagyClient(autoFallback: true);
    final events = <TtagyStreamEvent>[];

    await for (final ev in client.streamChat(const TtagyRequest(
      prompt: 'scenario:quota_error',
      model: 'gemini-3.7-flash',
      effort: 'low',
    ))) {
      events.add(ev);
    }

    expect(events.length, equals(2));
    expect(events[0], isA<TtagyInitEvent>());
    expect(events[1], isA<TtagyErrorEvent>());
    final err = events[1] as TtagyErrorEvent;
    expect(err.errorMessage, contains('Resource quota exceeded'));
  });
}

File: packages/ttagy-client/src/__tests__/conformance.test.mjs
import test from "node:test";
import assert from "node:assert/strict";
import * as path from "node:path";
import * as url from "node:url";
import { TtagyClient } from "../index.ts";

const __dirname = path.dirname(url.fileURLToPath(import.meta.url));
const mockAgyPath = path.resolve(__dirname, "../../../../target/debug/mock-agy");

test("TS CTS: stream_normal parity", async () => {
  process.env.AGY_PATH = mockAgyPath;

  const client = new TtagyClient({ autoFallback: true });
  const events = [];
  let content = "";

  for await (const ev of client.streamChat({
    prompt: "scenario:stream_normal",
    model: "gemini-3.7-flash",
    effort: "low",
  })) {
    events.push(ev);
    if (ev.type === "agy:content_delta") {
      content += ev.textDelta;
    }
  }

  const types = events.map((e) => e.type);
  assert.deepEqual(types, [
    "agy:init",
    "agy:thinking_delta",
    "agy:content_delta",
    "agy:content_delta",
    "agy:done",
  ]);
  assert.equal(content, "你好，我是 Antigravity AI 助手。很高兴为您服务！");
});

test("TS CTS: structured_json parity", async () => {
  process.env.AGY_PATH = mockAgyPath;

  const client = new TtagyClient({ autoFallback: true });
  const result = await client.runJson({
    prompt: "scenario:structured_json",
    model: "gemini-3.7-flash",
    effort: "low",
  });

  assert.equal(result.status, "success");
  assert.equal(result.task, "compression");
  assert.equal(result.files_count, 3);
});

test("TS CTS: quota_error parity", async () => {
  process.env.AGY_PATH = mockAgyPath;

  const client = new TtagyClient({ autoFallback: true });
  const events = [];

  for await (const ev of client.streamChat({
    prompt: "scenario:quota_error",
    model: "gemini-3.7-flash",
    effort: "low",
  })) {
    events.push(ev);
  }

  assert.equal(events.length, 2);
  assert.equal(events[0].type, "agy:init");
  assert.equal(events[1].type, "agy:error");
  assert.ok(events[1].errorMessage.includes("Resource quota exceeded"));
});
